message/validation/common_checks.go

```diff
@@ -13,70 +14,108 @@ func (mv *messageValidator) committeeRole(role spectypes.RunnerRole) bool {
+		ttl = phase0.Slot(mv.netCfg.Beacon.SlotsPerEpoch()) + lateSlotAllowance
 	case spectypes.RoleValidatorRegistration, spectypes.RoleVoluntaryExit:
 		return 0
 	}
 
 	deadline := mv.netCfg.Beacon.GetSlotStartTime(slot + ttl).
-		Add(lateMessageMargin).Add(clockErrorTolerance)
+		Add(lateMessageMargin)
 
-	return mv.netCfg.Beacon.GetSlotStartTime(mv.netCfg.Beacon.EstimatedSlotAtTime(receivedAt.Unix())).
-		Sub(deadline)
+	return receivedAt.Sub(deadline)
 }
 
 func (mv *messageValidator) validateDutyCount(
-	validatorIndices []phase0.ValidatorIndex,
-	state *SignerState,
 	msgID spectypes.MessageID,
-	newDutyInSameEpoch bool,
+	msgSlot phase0.Slot,
+	validatorIndices []phase0.ValidatorIndex,
+	signerStateBySlot *treemap.Map,
 ) error {
-	var dutyLimit int
+	msgEpoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot)
+	dutyCount := 0
+	signerStateBySlot.Each(func(slot any, state any) {
+		if mv.netCfg.Beacon.EstimatedEpochAtSlot(slot.(phase0.Slot)) == msgEpoch {
+			dutyCount++
+		}
+	})
+	if _, ok := signerStateBySlot.Get(msgSlot); !ok {
+		dutyCount++
+	}
 
+	dutyLimit, exists := mv.dutyLimit(msgID, validatorIndices)
+	if !exists {
+		return nil
+	}
+
+	if dutyCount > dutyLimit {
+		err := ErrTooManyDutiesPerEpoch
+		err.got = fmt.Sprintf("%v (role %v)", dutyCount, msgID.GetRoleType())
+		err.want = fmt.Sprintf("less than %v", dutyLimit)
+		return err
+	}
+
+	return nil
+}
+
+func (mv *messageValidator) dutyLimit(msgID spectypes.MessageID, validatorIndices []phase0.ValidatorIndex) (int, bool) {
 	switch msgID.GetRoleType() {
 	case spectypes.RoleAggregator, spectypes.RoleValidatorRegistration, spectypes.RoleVoluntaryExit:
-		dutyLimit = 2
+		return 2, true
 
 	case spectypes.RoleCommittee:
-		dutyLimit = 2 * len(validatorIndices)
+		return 2 * len(validatorIndices), true
 
 	default:
-		return nil
+		return 0, false
 	}
+}
 
-	if sameSlot := !newDutyInSameEpoch; sameSlot {
-		dutyLimit++
+func (mv *messageValidator) validateBeaconDuty(
+	role spectypes.RunnerRole,
+	slot phase0.Slot,
+	indices []phase0.ValidatorIndex,
+) error {
+	// Rule: For a proposal duty message, we check if the validator is assigned to it
+	if role == spectypes.RoleProposer {
+		epoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(slot)
+		if mv.dutyStore != nil && mv.dutyStore.Proposer.ValidatorDuty(epoch, slot, indices[0]) == nil {
+			return ErrNoDuty
+		}
 	}
 
-	if state.EpochDuties >= dutyLimit {
-		err := ErrTooManyDutiesPerEpoch
-		err.got = fmt.Sprintf("%v (role %v)", state.EpochDuties, msgID.GetRoleType())
-		err.want = fmt.Sprintf("less than %v", dutyLimit)
-		return err
+	// Rule: For a sync committee aggregation duty message, we check if the validator is assigned to it
+	if role == spectypes.RoleSyncCommitteeContribution {
+		period := mv.netCfg.Beacon.EstimatedSyncCommitteePeriodAtEpoch(mv.netCfg.Beacon.EstimatedEpochAtSlot(slot))
+		if mv.dutyStore != nil && mv.dutyStore.SyncCommittee.Duty(period, indices[0]) == nil {
+			return ErrNoDuty
+		}
 	}
 
```

Comment:
if `mv.dutyStore` doesn't exist this function will return nil which IMO isn't correct, we should return a separate error for example dutyStore doesn't exist and on the upper level decide if to ignore and just go on or do something else.

Reply:
@y0sher We never pass nil duty store to the message validator constructor, this is actually a leftover from the previous message validation when the duty store was optional. So I think we can just remove the check and panic if it's nil

Reply:
@y0sher done